app/src/main/java/com/example/login/canvas/TicTacBoard.kt

```diff
@@ -0,0 +1,108 @@
+package com.example.login.canvas
+
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.RectF
+import android.util.AttributeSet
+import android.widget.GridLayout
+import androidx.core.content.res.ResourcesCompat
+import com.example.login.R
+
+class TicTacBoard @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : GridLayout(context, attrs, defStyleAttr) {
+
+    private var mWinLineState: WinLineState = WinLineState.None
+    private var winLineRect: RectF? = null
+
+    private var widthF: Float = 0f
+    private var heightF: Float = 0f
+    private var cellWidth: Float = 0f
+        set(value) {
+            field = value
+            halfCellWidth = value / 2
+        }
+
+    private var cellHeight: Float = 0f
+        set(value) {
+            field = value
+            halfCellHeight = value / 2
+        }
+    private var halfCellWidth: Float = 0f
+
+    private var halfCellHeight: Float = 0f
+    private val strokeSize = resources.getDimension(R.dimen.stroke_size)
+
+    private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
+        color = ResourcesCompat.getColor(context.resources, R.color.item_color, null)
+        strokeWidth = strokeSize
+        strokeCap = Paint.Cap.ROUND
+    }
+
+    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
+        super.onMeasure(widthSpec, heightSpec)
+        widthF = MeasureSpec.getSize(widthSpec).toFloat()
+        heightF = MeasureSpec.getSize(heightSpec).toFloat()
+
+        cellWidth = widthF / 3f
```

Comment:
3f should be getting through Game class that provides all definition of game include board dimension

Reply:
done +-